flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/DateTimeUtils.java

```diff
@@ -412,92 +492,267 @@ public static Long toTimestamp(String dateStr, TimeZone tz) {
      */
-    public static String dateFormat(long ts, String format, TimeZone tz) {
-        SimpleDateFormat formatter = FORMATTER_CACHE.get(format);
-        formatter.setTimeZone(tz);
-        Date dateTime = new Date(ts);
-        return formatter.format(dateTime);
-    }
-
+    private static int parseFraction(String v) {
+        int multiplier = 100;
+        int r = 0;
+        for (int i = 0; i < v.length(); i++) {
+            char c = v.charAt(i);
+            int x = c < '0' || c > '9' ? 0 : (c - '0');
+            r += multiplier * x;
+            if (multiplier < 10) {
+                // We're at the last digit. Check for rounding.
+                if (i + 1 < v.length() && v.charAt(i + 1) >= '5') {
+                    ++r;
+                }
+                break;
+            }
+            multiplier /= 10;
+        }
+        return r;
+    }
+
+    // --------------------------------------------------------------------------------------------
+    // Format
+    // --------------------------------------------------------------------------------------------
+
+    public static String formatTimestamp(TimestampData ts, String format) {
+        return formatTimestamp(ts, format, ZoneId.of("UTC"));
+    }
+
+    public static String formatTimestamp(TimestampData ts, String format, TimeZone zone) {
+        return formatTimestamp(ts, format, zone.toZoneId());
+    }
+
+    private static String formatTimestamp(TimestampData ts, int precision) {
+        LocalDateTime ldt = ts.toLocalDateTime();
+
+        String fraction = pad(9, ldt.getNano());
+        while (fraction.length() > precision && fraction.endsWith("0")) {
+            fraction = fraction.substring(0, fraction.length() - 1);
+        }
+
+        StringBuilder ymdhms =
+                ymdhms(
+                        new StringBuilder(),
+                        ldt.getYear(),
+                        ldt.getMonthValue(),
+                        ldt.getDayOfMonth(),
+                        ldt.getHour(),
+                        ldt.getMinute(),
+                        ldt.getSecond());
+
+        if (fraction.length() > 0) {
+            ymdhms.append(".").append(fraction);
+        }
+
+        return ymdhms.toString();
+    }
+
+    public static String formatTimestamp(TimestampData ts, TimeZone tz, int precision) {
+        return formatTimestamp(timestampWithLocalZoneToTimestamp(ts, tz), precision);
+    }
+
+    /**
+     * Format a timestamp as specific.
+     *
+     * @param ts the {@link TimestampData} to format.
+     * @param format the string formatter.
+     * @param zoneId the ZoneId.
+     */
+    private static String formatTimestamp(TimestampData ts, String format, ZoneId zoneId) {
+        DateTimeFormatter formatter = DATETIME_FORMATTER_CACHE.get(format);
+        Instant instant = ts.toInstant();
+        return LocalDateTime.ofInstant(instant, zoneId).format(formatter);
+    }
+
+    /**
+     * Format a timestamp as specific.
```

Comment:
```suggestion
     * Format a timestamp as string using a custom format.
```
Although I don't think that javadoc is usuful here, the method is quite obvious by its name and args, and it's a private one.

Reply:
Removed javadoc